diff --git a/fastboot/Android.bp b/fastboot/Android.bp
index 716fe95..20fa9f4 100644
--- a/fastboot/Android.bp
+++ b/fastboot/Android.bp
@@ -208,7 +208,7 @@ cc_defaults {
 //
 
 cc_library_host_static {
-    name: "libfastboot",
+    name: "libhfastboot",
     defaults: ["fastboot_host_defaults"],
 
     srcs: [
@@ -248,11 +248,11 @@ cc_library_host_static {
 //
 
 cc_binary_host {
-    name: "fastboot",
+    name: "hfastboot",
     defaults: ["fastboot_host_defaults"],
 
     srcs: ["main.cpp"],
-    static_libs: ["libfastboot"],
+    static_libs: ["libhfastboot"],
 
     required: [
         "mke2fs",
@@ -307,5 +307,86 @@ cc_test_host {
             // Avoid trying to build for win64
             enabled: false,
         },
+        windows_x86: {
+            // Avoid trying to build for win32
+            enabled: false,
+        },
+        linux: {
+            // Prevent error: missing variant: arch:linux_glibc_x86_64
+            enabled: false,
+        },
     },
 }
+
+//
+// Build fastboot for Android
+//
+
+cc_defaults {
+    name: "fastboot_android_defaults",
+
+    use_version_lib: true,
+
+    cflags: [
+        "-Wall",
+        "-Wextra",
+        "-Werror",
+        "-Wunreachable-code",
+        "-Wno-non-virtual-dtor",
+    ],
+
+    stl: "libc++_static",
+
+    // Don't add anything here, we don't want additional shared dependencies
+    // on the host fastboot tool, and shared libraries that link against libc++
+    // will violate ODR.
+    shared_libs: ["libcrypto"],
+
+    header_libs: ["bootimg_headers"],
+
+    static_libs: [
+        "libziparchive",
+        "libsparse",
+        "libutils",
+        "liblog",
+        "libz",
+        "libdiagnose_usb",
+        "libbase",
+        "libcutils",
+        "liblp",
+        "libext4_utils",
+    ],
+}
+
+cc_library_static {
+    name: "libfastboot",
+    defaults: ["fastboot_android_defaults"],
+    recovery: true,
+
+    srcs: [
+        "bootimg_utils.cpp",
+        "fastboot.cpp",
+        "fs.cpp",
+        "socket.cpp",
+        "tcp.cpp",
+        "udp.cpp",
+        "util.cpp",
+        "fastboot_driver.cpp",
+        "usb_linux.cpp",
+    ],
+
+    // Only version the final binaries
+    use_version_lib: false,
+    static_libs: ["libbuildversion"],
+
+    generated_headers: ["platform_tools_version"],
+}
+
+cc_binary {
+    name: "fastboot",
+    defaults: ["fastboot_android_defaults"],
+    recovery: true,
+
+    srcs: ["main.cpp"],
+    static_libs: ["libfastboot"],
+}
diff --git a/fastboot/socket.h b/fastboot/socket.h
index e791f2c..5200462 100644
--- a/fastboot/socket.h
+++ b/fastboot/socket.h
@@ -40,7 +40,6 @@
 
 #include <android-base/macros.h>
 #include <cutils/sockets.h>
-#include <gtest/gtest_prod.h>
 
 // Socket interface to be implemented for each platform.
 class Socket {
@@ -119,8 +118,5 @@ class Socket {
             socket_send_buffers_function_ = &socket_send_buffers;
 
   private:
-    FRIEND_TEST(SocketTest, TestTcpSendBuffers);
-    FRIEND_TEST(SocketTest, TestUdpSendBuffers);
-
     DISALLOW_COPY_AND_ASSIGN(Socket);
 };
