diff --git a/init/property_service.cpp b/init/property_service.cpp
index 53229eb..662937a 100644
--- a/init/property_service.cpp
+++ b/init/property_service.cpp
@@ -683,6 +683,20 @@ static void update_sys_usb_config() {
 }
 
 void property_load_boot_defaults() {
+    std::string xiaomi_device;
+    if (ReadFileToString("/proc/device-tree/xiaomi,device", &xiaomi_device)) {
+        property_set("ro.build.product", xiaomi_device);
+        property_set("ro.product.device", xiaomi_device);
+        property_set("ro.product.odm.device", xiaomi_device);
+        property_set("ro.product.system.device", xiaomi_device);
+        property_set("ro.product.vendor.device", xiaomi_device);
+    }
+
+    // Force ADB on boot
+    property_set("ro.secure","0");
+    property_set("ro.adb.secure","0");
+    property_set("ro.debuggable","1");
+
     if (!load_properties_from_file("/system/etc/prop.default", NULL)) {
         // Try recovery path
         if (!load_properties_from_file("/prop.default", NULL)) {
@@ -697,6 +711,9 @@ void property_load_boot_defaults() {
     if (android::base::GetBoolProperty("ro.persistent_properties.ready", false)) {
         update_sys_usb_config();
     }
+
+    // Force ADB on boot
+    property_set("persist.sys.usb.config","adb");
 }
 
 static void load_override_properties() {
