From c1e47423ee1d4bde1967f037ba4bb2fda0b71167 Mon Sep 17 00:00:00 2001
From: iusmac <iusico.maxim@libero.it>
Date: Thu, 16 Jun 2022 20:24:35 +0200
Subject: [PATCH] data: Concatenate battery temperature with CPU temperature

Signed-off-by: iusmac <iusico.maxim@libero.it>
---
 data.cpp | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/data.cpp b/data.cpp
index ef52a34..b5f7812 100755
--- a/data.cpp
+++ b/data.cpp
@@ -1459,11 +1459,14 @@ int DataManager::GetMagicValue(const string& varName, string& value)
 		GetValue("tw_no_cpu_temp", tw_no_cpu_temp);
 		if (tw_no_cpu_temp == 1) return -1;
 
+		char tmp[19];
 		string cpu_temp_file;
 		static unsigned long convert_temp = 0;
+		static unsigned long convert_temp_batt = 0;
 		static time_t cpuSecCheck = 0;
 		struct timeval curTime;
 		string results;
+		string results_batt;
 
 		gettimeofday(&curTime, NULL);
 		if (curTime.tv_sec > cpuSecCheck)
@@ -1482,9 +1485,14 @@ int DataManager::GetMagicValue(const string& varName, string& value)
 				convert_temp = strtoul(results.c_str(), NULL, 0);
 			if (convert_temp >= 150)
 				convert_temp = strtoul(results.c_str(), NULL, 0) / 10;
+
+			if (TWFunc::read_file("/sys/class/power_supply/battery/temp", results_batt) != 0) return -1;
+			convert_temp_batt = strtoul(results_batt.c_str(), NULL, 0) / 10;
+
 			cpuSecCheck = curTime.tv_sec + 5;
 		}
-		value = TWFunc::to_string(convert_temp);
+		sprintf(tmp, "%lu °C | BATT: %lu", convert_temp, convert_temp_batt);
+		value = tmp;
 		return 0;
 	}
 	return -1;

-- 
2.32.0
