diff --git a/twrp-functions.cpp b/twrp-functions.cpp
index 4f4d88e..f5d1574 100755
--- a/twrp-functions.cpp
+++ b/twrp-functions.cpp
@@ -2382,7 +2382,7 @@ void TWFunc::OrangeFox_Startup(void)
   cpu_two = "/cpufreq/scaling_governor";
   std::string enable = "1";
   std::string disable = "0";
-  std::string t2w = "/sys/android_touch/doubletap2wake";
+  std::string t2w = "/dev/null";
   std::string fsync = "/sys/module/sync/parameters/fsync_enabled";
   std::string fast_charge = "/sys/kernel/fast_charge/force_fast_charge";
   std::string performance = "performance";
