diff --git a/partitionmanager.cpp b/partitionmanager.cpp
index 00b883c67..daeb8fc3b 100755
--- a/partitionmanager.cpp
+++ b/partitionmanager.cpp
@@ -3667,6 +3667,7 @@ void TWPartitionManager::Translate_Partition_Display_Names()
 }
 
 bool TWPartitionManager::Decrypt_Adopted() {
+/*
 #ifdef TW_INCLUDE_CRYPTO
 
 	int SDK = TWFunc::Get_Android_SDK_Version();
@@ -3780,9 +3781,10 @@ bool TWPartitionManager::Decrypt_Adopted() {
 	}
 	return ret;
 #else
-	LOGINFO("Decrypt_Adopted: no crypto support\n");
+*/
+	LOGINFO("Decrypt_Adopted: not supported on Android 12\n");
 	return false;
-#endif
+//#endif
 }
 
 void TWPartitionManager::Remove_Partition_By_Path(string Path)
-- 
2.32.0

diff --git a/crypto/fde/cryptfs.cpp b/crypto/fde/cryptfs.cpp
index 431e71e54..f5b6b3b3f 100644
--- a/crypto/fde/cryptfs.cpp
+++ b/crypto/fde/cryptfs.cpp
@@ -116,6 +116,7 @@ constexpr size_t INTERMEDIATE_BUF_SIZE =
 static_assert(INTERMEDIATE_BUF_SIZE == SCRYPT_LEN,
               "Mismatch of intermediate key sizes");
 
+static const std::string kPkmBlob("pKMblob\x00", 8);
 #define KEY_IN_FOOTER  "footer"
 
 #define DEFAULT_HEX_PASSWORD "64656661756c745f70617373776f7264"
@@ -440,6 +441,14 @@ static int keymaster_sign_object(struct crypt_mnt_ftr *ftr,
 
     int rc = -1;
 
+    // In A12 keymaster_key_blob format changed:
+    // it have useless for us bytes in beginning, so remove them to correctly handle key
+    std::string kmKey( &ftr->keymaster_blob[0], &ftr->keymaster_blob[0]+ftr->keymaster_blob_size );
+    if (!kmKey.compare(0, kPkmBlob.size(), kPkmBlob)) {
+        kmKey.erase(0, kPkmBlob.size());
+    }
+    keymaster_key_blob_t key = { reinterpret_cast<const uint8_t*>(kmKey.c_str()), kmKey.size() };
+
 #if TW_KEYMASTER_MAX_API >= 1
     keymaster0_device_t *keymaster0_dev = 0;
     keymaster1_device_t *keymaster1_dev = 0;
@@ -463,8 +472,8 @@ static int keymaster_sign_object(struct crypt_mnt_ftr *ftr,
 
         rc = keymaster0_dev->sign_data(keymaster0_dev,
                                       &params,
-                                      ftr->keymaster_blob,
-                                      ftr->keymaster_blob_size,
+                                      key.key_material,
+                                      key.key_material_size,
                                       to_sign,
                                       to_sign_size,
                                       signature,
@@ -473,7 +482,6 @@ static int keymaster_sign_object(struct crypt_mnt_ftr *ftr,
     }
 #if TW_KEYMASTER_MAX_API >= 1
     else if (keymaster1_dev) {
-        keymaster_key_blob_t key = { ftr->keymaster_blob, ftr->keymaster_blob_size };
         keymaster_key_param_t params[] = {
             keymaster_param_enum(KM_TAG_PADDING, KM_PAD_NONE),
             keymaster_param_enum(KM_TAG_DIGEST, KM_DIGEST_NONE),
@@ -530,7 +538,6 @@ static int keymaster_sign_object(struct crypt_mnt_ftr *ftr,
     }
     #ifndef OF_DISABLE_KEYMASTER2
     else if (keymaster2_dev) {
-        keymaster_key_blob_t key = { ftr->keymaster_blob, ftr->keymaster_blob_size };
         keymaster_key_param_t params[] = {
             keymaster_param_enum(KM_TAG_PADDING, KM_PAD_NONE),
             keymaster_param_enum(KM_TAG_DIGEST, KM_DIGEST_NONE),
-- 
2.32.0

diff --git a/crypto/ext4crypt/Keymaster3.h b/crypto/ext4crypt/Keymaster3.h
index 521571ba8..3820f1b28 100644
--- a/crypto/ext4crypt/Keymaster3.h
+++ b/crypto/ext4crypt/Keymaster3.h
@@ -111,6 +111,34 @@ class Keymaster {
     DISALLOW_COPY_AND_ASSIGN(Keymaster);
 };
 
+namespace dump {
+
+template<typename T>
+std::string toHexString(T t, bool prefix = true) {
+    std::ostringstream os;
+    if (prefix) { os << std::showbase; }
+    os << std::hex << t;
+    return os.str();
+}
+
+template<typename Array>
+std::string arrayToHexString(const Array &a, size_t size) {
+    using android::hardware::toString;
+    std::string os;
+    for (size_t i = 0; i < size; ++i) {
+        os += toHexString(a[i]);
+    }
+    return os;
+}
+
+template<typename T>
+std::string toString(const hardware::hidl_vec<T> &a) {
+    std::string os;
+    os += arrayToHexString(a, a.size());
+    return os;
+}
+
+}  // namespace dump
 }  // namespace vold
 }  // namespace android
 
-- 
2.32.0

diff --git a/crypto/ext4crypt/Keymaster3.cpp b/crypto/ext4crypt/Keymaster3.cpp
index f0e578de1..49adeaccc 100644
--- a/crypto/ext4crypt/Keymaster3.cpp
+++ b/crypto/ext4crypt/Keymaster3.cpp
@@ -25,6 +25,8 @@
 #define ERROR 1
 #define LOG(x) std::cout
 
+static const std::string kPkmBlob("pKMblob\x00", 8);
+
 using namespace ::keystore;
 using android::hardware::hidl_string;
 
@@ -179,6 +181,14 @@ KeymasterOperation Keymaster::begin(KeyPurpose purpose, const std::string& key,
         mOpHandle = operationHandle;
     };
 
+    // In A12 keymaster_key_blob format changed:
+    // it have useless for us bytes in beginning, so remove them to correctly handle key
+    std::string kmKey = dump::toString(keyBlob);
+    if (!kmKey.compare(0, kPkmBlob.size(), kPkmBlob)) {
+        kmKey.erase(0, kPkmBlob.size());
+        keyBlob = blob2hidlVec(kmKey);
+    }
+
     auto error = mDevice->begin(purpose, keyBlob, inParams.hidl_data(), hidlCb);
     if (!error.isOk()) {
         LOG(ERROR) << "begin failed: " << error.description();
-- 
2.32.0

