diff --git a/FoxExtras/sbin/foxstart.sh b/FoxExtras/sbin/foxstart.sh
index eaddec6..e5b9fc3 100755
--- a/FoxExtras/sbin/foxstart.sh
+++ b/FoxExtras/sbin/foxstart.sh
@@ -427,22 +427,8 @@ backup_restore_FS() {
 
 # fix yellow flashlight on mido/vince/kenzo and configure Leds on others
 flashlight_Leds_config() {
-   case "$FOX_DEVICE" in
-       vince)
-   		echo "0" > /sys/devices/soc/qpnp-flash-led-24/leds/led:torch_1/max_brightness;
-       	;;
-       riva)
-            	echo 1 > /sys/class/leds/flashlight/max_brightness;
-            	echo 0 > /sys/class/leds/flashlight/brightness;
-         ;;
-       *)
-       		return;
-       	;;
-   esac
-
-   echo "0" > /sys/class/leds/led:torch_1/max_brightness
-   echo "0" > /sys/class/leds/torch-light1/max_brightness
-   echo "0" > /sys/class/leds/led:flash_1/max_brightness
+   echo 1 > /sys/class/leds/flashlight/max_brightness;
+   echo 0 > /sys/class/leds/flashlight/brightness;
 }
 
 # start, and mark that we have started
