diff --git a/installer/META-INF/com/google/android/update-binary b/installer/META-INF/com/google/android/update-binary
index 4f4e8b9..7a38f02 100644
--- a/installer/META-INF/com/google/android/update-binary
+++ b/installer/META-INF/com/google/android/update-binary
@@ -218,6 +218,7 @@ local F=/prop.default
 CheckRequirements() {
    local GREP=$(find_grep)
    local F=$(echo "$PRODUCT_DEVICE" | $GREP -w "$TARGET_DEVICE")
+   return
    if [ -n "$F" ]; then
        ui_print "- OrangeFox (RELEASE_VER) for $TARGET_DEVICE ..."
    else
